### PR TITLE
feature/wallet: fetch transactions for all zones

### PR DESCRIPTION
**Pull request by hsyndeniz**
### Description

Fetching Transactions from Blockscout: The code has been modified to enable fetching transactions from Blockscout for all shards.

Displaying Balances for All Zones: The user interface has been updated to include balance information for all zones. With this improvement, users can easily monitor their wallet balances across different shards.

### CODE REVIEW

File: src/shared/components/QuaiPayActiveAddressModal.tsx [hunk 1]
```diff
@@ -78,15 +78,20 @@ const Modal: ForwardRefRenderFunction<
                           : styles.textNotSelected
                       }
                     >
-                      {_.balances[walletObject![zone].address] || '0.00'} Quai
+                      {(_.balances &&
```
Comment (juanmanso) [inline src/shared/components/QuaiPayActiveAddressModal.tsx:81]: could we do destructuring here instead of indexing a `_` variable (which is usually called like this because it's not used)

File: src/shared/components/QuaiPayActiveAddressModal.tsx [hunk 2]
```diff
@@ -78,15 +78,20 @@ const Modal: ForwardRefRenderFunction<
...
+                        _.balances[walletObject![zone].address]) ||
```
Comment (juanmanso) [inline src/shared/components/QuaiPayActiveAddressModal.tsx:82]: any reason why you are doing the non-null assertion here? Is it because `walletObject` could be `undefined`?

If that's so, I would suggest to do the explicit check for `walletObject` because the user might refuse entering their keychain passcode and I'm afraid it would break the app 🤔

File: src/main/wallet/WalletScreen.tsx [hunk 3]
```diff
@@ -69,71 +69,175 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
...
+              return Number(item.timeStamp) >= ninetyDaysAgo.getTime() / 1000;
+            } else if (selectedTimeframe === 'Past 30 days') {
+              return Number(item.timeStamp) >= thirtyDaysAgo.getTime() / 1000;
+            } else if (selectedTimeframe === 'This week') {
+              return Number(item.timeStamp) >= sevenDaysAgo.getTime() / 1000;
+            }
+          }
+          return true;
+        });
+        const mappedTransactions = filteredTransactions.map(item => {
```
Comment (juanmanso) [inline src/main/wallet/WalletScreen.tsx:191]: same here. The function itself it's readable but cluttered with everything else makes it hard to follow. Maybe we could extract the callback so the high-level of the flow is easier to understand, what do you think? Or you feel that might be some overkill?

File: src/main/wallet/FilterModal.tsx [hunk 4]
```diff
@@ -53,7 +53,7 @@ export const FilterModal = forwardRef<BottomSheetModal, FilterModalProps>(
       0, 1, 2, 3, 4, 5, 6, 7, 8,
     ]);
 
-    const txDirection = useMemo(() => [t('from'), t('toDirection')], []);
+    const txDirection = useMemo(() => ['from', 'to'], []);
```
Comment (juanmanso) [inline src/main/wallet/FilterModal.tsx:56]: any reason why we are disabling i18n here? Am I missing something?

File: src/shared/components/QuaiPayActiveAddressModal.tsx [hunk 5]
```diff
@@ -9,60 +9,51 @@ import { abbreviateAddress } from 'src/shared/services/quais';
...
-                  key={zone}
-                  onPress={() => {
-                    if (!isSelected) {
-                      setZone(zone);
-                    }
-                  }}
+const Modal: ForwardRefRenderFunction<
+  BottomSheetModal,
+  QuaiPayActiveAddressModalProps
+> = (_, ref) => {
```
Comment (juanmanso) [inline src/shared/components/QuaiPayActiveAddressModal.tsx:23]: any reason to redefine the component?

File: src/shared/components/QuaiPayActiveAddressModal.tsx [hunk 6]
```diff
@@ -9,60 +9,51 @@ import { abbreviateAddress } from 'src/shared/services/quais';
...
+  const { zone: selectedZone, setZone, walletObject } = useWalletContext();
+  const styles = useThemedStyle(themedStyle);
+
+  return (
+    <QuaiPayBottomSheetModal ref={ref}>
+      <QuaiPayText style={styles.title}>Choose Active Address</QuaiPayText>
```
Comment (juanmanso) [inline src/shared/components/QuaiPayActiveAddressModal.tsx:29]: could we use i18n here?

**Review by juanmanso — CHANGES_REQUESTED** [review]
Left you some comments, let me know if I can help you clarify or with anything else

File: src/shared/components/QuaiPayActiveAddressModal.tsx (hunk 2 above)
Comment (obasilakis) [reply src/shared/components/QuaiPayActiveAddressModal.tsx:82]: we should always show a loader when `walletObject` is undefined

File: src/main/wallet/WalletScreen.tsx [hunk 7]
```diff
@@ -69,71 +69,171 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
...
+          const address =
+            (walletObject && walletObject[key as Zone].address) || '';
+          try {
+            const res = await getBalance(address, zone);
+            const formattedBalance = Number(
+              quais.utils.formatEther(res),
+            ).toFixed(3);
+            return { address: address, balance: formattedBalance };
+          } catch (error) {
+            return { address: address, balance: 'Error' };
```
Comment (obasilakis) [inline src/main/wallet/WalletScreen.tsx:87]: we should show a banner with the error and tell them to retry later instead of setting the balance to the word `Error` here, otherwise you get this:
Simulator Screenshot - iPhone 14 - 2023-07-18 at 21 39 37

File: src/main/wallet/WalletScreen.tsx [hunk 8]
```diff
@@ -69,71 +69,171 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
...
+      const zoneShard = Object.keys(walletObject as object)[shard];
+      const address =
+        walletObject?.[zoneShard as keyof typeof walletObject]?.address;
+      try {
+        const res = await getAccountTransactions(
+          {
+            address: address || '',
+            sort: 'desc',
+            page: 1,
+            offset: 100,
```
Comment (obasilakis) [inline src/main/wallet/WalletScreen.tsx:140]: let's dial that down to 30 and take care of infinite scrolling

File: src/main/wallet/FilterModal.tsx (hunk 4 above)
Comment (juanmanso) [reply src/main/wallet/FilterModal.tsx:56]: cool, resolving this comment then. Thanks for responding

File: src/shared/services/dateUtil.ts [hunk 9]
```diff
@@ -11,3 +11,38 @@ export const dateToLocaleString = (date: Date) => {
...
+    'Past 90 days': getTimeframeDate(90),
+    'Past 30 days': getTimeframeDate(30),
+    'This week': getTimeframeDate(7),
+  };
+
+  const selectedTimestamp = timeframes[selectedTimeframe];
+
+  if (selectedTimestamp === undefined) {
+    throw new Error('Invalid timeframe selected.');
+  }
```
Comment (juanmanso) [inline src/shared/services/dateUtil.ts:45]: Since you already typed `selectedTimeframe` as `Timeframe` it cannot be undefined and this won't happen, is that right?

File: src/shared/components/QuaiPayActiveAddressModal.tsx [hunk 10]
```diff
@@ -20,13 +21,15 @@ interface QuaiPayActiveAddressModalProps {
 const Modal: ForwardRefRenderFunction<
   BottomSheetModal,
   QuaiPayActiveAddressModalProps
-> = (_, ref) => {
+> = (params, ref) => {
```
Comment (juanmanso) [inline src/shared/components/QuaiPayActiveAddressModal.tsx:24]: nitpick improvement: you can destruct here and get `balances` directly without the intermediate `params`

File: src/shared/components/QuaiPayActiveAddressModal.tsx [hunk 11]
```diff
@@ -20,13 +21,15 @@ interface QuaiPayActiveAddressModalProps {
...
+  const { balances } = params;
   const { zone: selectedZone, setZone, walletObject } = useWalletContext();
   const styles = useThemedStyle(themedStyle);
+  const { t } = useTranslation('translation', { keyPrefix: 'wallet' });
 
   return (
     <QuaiPayBottomSheetModal ref={ref}>
-      <QuaiPayText style={styles.title}>Choose Active Address</QuaiPayText>
+      <QuaiPayText style={styles.title}>{t('chooseAddress')}</QuaiPayText>
```
Comment (juanmanso) [inline src/shared/components/QuaiPayActiveAddressModal.tsx]: nitpick: this component is shared across the app, I don't think we should prefix this localization with `wallet` but rather use the name of the component. What do you think about it

File: src/main/wallet/WalletScreen.tsx [hunk 12]
```diff
@@ -49,17 +54,16 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
...
   >();
   const [selectedTimeframe, setSelectedTimeframe] = useState<
-    string | undefined
+    Timeframe | undefined
   >();
-  const [minAmount, setMinAmount] = useState(1);
-  const [maxAmount, setMaxAmount] = useState(1000000000000000000000000);
-  const [shards, setShards] = useState<number[]>([]);
+  const [minAmount, setMinAmount] = useState(0);
+  const [maxAmount, setMaxAmount] = useState(1000000000000000000);
```
Comment (juanmanso) [inline src/main/wallet/WalletScreen.tsx:60]: nitpick: could we use scientific notation here to improve readability? I have to count the zeros and it's a bit annoying 😅. Using something like `1e12` or something would be great.

File: src/main/wallet/WalletScreen.tsx [hunk 13]
```diff
@@ -340,7 +329,7 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
                 quaiAmount={parseFloat(item.quaiAmount.toFixed(6)).toString()}
               />
             )}
-            keyExtractor={item => item.hash}
+            keyExtractor={item => Math.random().toString()}
```
Comment (juanmanso) [inline src/main/wallet/WalletScreen.tsx:332]: could we use something like the index here? I know it's unlikely but what if 2 keys are the same? This isn't deterministic sadly

File: src/main/wallet/WalletScreen.tsx [hunk 14]
```diff
@@ -304,23 +289,27 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
...
         {transactions.length === 0 ? (
-          <QuaiPayText type="paragraph">{t('noTransaction')}</QuaiPayText>
+          <QuaiPayText type="paragraph">
+            {t('wallet.noTransaction')}
+          </QuaiPayText>
         ) : (
           <FlatList
             data={transactions}
             contentContainerStyle={styles.flatlistContainer}
+            onEndReached={() => console.log('end reached')}
```
Comment (juanmanso) [inline src/main/wallet/WalletScreen.tsx:311]: can we remove this?

File: src/main/wallet/WalletScreen.tsx [hunk 15]
```diff
@@ -49,17 +54,16 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
...
-  const [maxAmount, setMaxAmount] = useState(1000000000000000000000000);
-  const [shards, setShards] = useState<number[]>([]);
+  const [minAmount, setMinAmount] = useState(0);
+  const [maxAmount, setMaxAmount] = useState(1000000000000000000);
+  const [shards, setShards] = useState<number[]>([0, 1, 2, 3, 4, 5, 6, 7, 8]);
   const [loading, setLoading] = useState(false);
-  // TODO: remove console.log when we use these values
-  console.log(shards, selectedTimeframe);
-
+  const [balances, setBalances] = useState<Record<string, string>>({});
```
Comment (juanmanso) [inline src/main/wallet/WalletScreen.tsx:63]: wouldn't it be better to store this with the zone as key? Also I think that `Record<string, string>` doesn't add value to the obj usage and definition (because it's how it normally works).

Let me know if I'm missing something

File: src/main/wallet/WalletScreen.tsx [hunk 16]
```diff
@@ -49,17 +54,16 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
     string | undefined
   >();
   const [selectedTimeframe, setSelectedTimeframe] = useState<
-    string | undefined
+    Timeframe | undefined
```
Comment (juanmanso) [inline src/main/wallet/WalletScreen.tsx:57]: you don't need to type this with `undefined` because you are already defining it as undefined in the parenthesis.

Just type it with `Timeframe` and it should work

File: src/main/wallet/WalletScreen.tsx [hunk 17]
```diff
@@ -106,38 +113,23 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
...
 
   useEffect(() => {
     setLoading(true);
     const txAllShards: Transaction[] = [];
+
     const promises = shards.map(async shard => {
-      const zoneShard = Object.keys(walletObject as object)[shard];
-      const address =
-        walletObject?.[zoneShard as keyof typeof walletObject]?.address;
+      const zoneShard = Object.values(walletObject as object)[shard];
```
Comment (juanmanso) [inline src/main/wallet/WalletScreen.tsx:124]: instead of casting an undefined value to an object, wouldn't it be simpler to read and maintain to add the check against undefined at the beginning of the useEffect (by doing the `if(walletObject)`)? What do you think?

File: src/main/wallet/WalletScreen.tsx [hunk 18]
```diff
@@ -146,77 +138,66 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
...
-              return Number(item.timeStamp) >= oneYearAgo.getTime() / 1000;
-            } else if (selectedTimeframe === 'Past 6 months') {
-              return Number(item.timeStamp) >= sixMonthsAgo.getTime() / 1000;
-            } else if (selectedTimeframe === 'Past 90 days') {
-              return Number(item.timeStamp) >= ninetyDaysAgo.getTime() / 1000;
-            } else if (selectedTimeframe === 'Past 30 days') {
-              return Number(item.timeStamp) >= thirtyDaysAgo.getTime() / 1000;
-            } else if (selectedTimeframe === 'This week') {
-              return Number(item.timeStamp) >= sevenDaysAgo.getTime() / 1000;
+        const filteredTransactions = res.result
```
Comment (juanmanso) [inline src/main/wallet/WalletScreen.tsx:142]: by reading it again with the new cleaner version I'm thinking that you are filtering and parsing in the same line. Is that ok? Shouldn't we split them for more readability/maintainability or is it an overkill?

**Review by juanmanso — CHANGES_REQUESTED** [review]
Left you some more comments, please check linter and conflicts with `main` before reviewing it again

File: src/main/wallet/WalletScreen.tsx (hunk 13 above)
Comment (hsyndeniz) [reply src/main/wallet/WalletScreen.tsx:332]: it was for testing with another chain. I replaced it with .hash again

File: src/main/wallet/WalletScreen.tsx [hunk 19]
```diff
@@ -299,6 +407,9 @@ const themedStyle = (theme: Theme) =>
       backgroundColor: theme.surface,
       padding: 16,
     },
+    flatlistContainer: {
+      paddingBottom: 280,
```
[the remark this answers is not in the document]
Comment (obasilakis) [reply src/main/wallet/WalletScreen.tsx:387]: adding a TODO since I can't test while the faucet isn't working

File: src/main/wallet/WalletScreen.tsx [hunk 20]
```diff
@@ -70,72 +65,52 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
...
-              ? {
-                  display: contact.username,
-                  profilePicture: contact.profilePicture,
-                }
-              : {
-                  display: isUserSender
-                    ? abbreviateAddress(item.to)
-                    : abbreviateAddress(item.from),
-                };
+      showSnackBar,
```
Comment (juanmanso) [inline src/main/wallet/WalletScreen.tsx:76]: I think passing the `showSnackBar` function here is wrong because we are mixing UI with logic.

The best approach (at least in my opinion) would be to throw an error on the `updateBalances` function and catch it here in the component and in there use the `showSnackBar` fx

> _Let's leave it as it is for now and improve it later_

File: src/main/wallet/WalletScreen.tsx [hunk 21]
```diff
@@ -70,72 +65,52 @@ const WalletScreen: React.FC<MainTabStackScreenProps<'Wallet'>> = () => {
...
+      zone,
+      quaiRate,
+      wallet,
+      setTransactions,
+      transactions,
+      setLoading,
+      selectedTimeframe,
+      selectedTxDirection,
+      contacts,
+    );
```
Comment (juanmanso) [inline src/main/wallet/WalletScreen.tsx:103]: Don't like how this looks, too many params 😅 .

Let's take note and see if we can refactor it later (for example removing the UI logic related params).

File: src/shared/components/QuaiPayActiveAddressModal.tsx [hunk 22]
```diff
@@ -9,63 +9,62 @@ import { abbreviateAddress } from 'src/shared/services/quais';
...
-              const isSelected = selectedZone === zone;
-              return (
-                <Pressable
-                  key={zone}
-                  onPress={() => {
-                    if (!isSelected) {
-                      setZone(zone);
-                    }
-                  }}
+const Modal: ForwardRefRenderFunction<
```
Comment (juanmanso) [inline src/shared/components/QuaiPayActiveAddressModal.tsx:25]: I still think we shouldn't redefine it, but we can take a look at it later

**Review by juanmanso — APPROVED** [review]
There a couple of improvements to be made, but in general I think we are safe to merge 🚀
